nixos/nginx: remove recommendedZstdSettings, add experimental option

fpletz opened the pull request:
The zstd nginx module has known bugs and upstream is currently not maintained. We should not recommend a buggy module and configuration to our users since we are not maintaining the module either.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -1119,6 +1119,12 @@ in
     (mkRenamedOptionModule [ "services" "nginx" "proxyCache" "keysZoneName" ] [ "services" "nginx" "proxyCachePath" "" "keysZoneName" ])
     (mkRenamedOptionModule [ "services" "nginx" "proxyCache" "enable" ] [ "services" "nginx" "proxyCachePath" "" "enable" ])
+    (mkRemovedOptionModule [ "services" "nginx" "recommendedZstdSettings" ] ''
+      The Nginx zstd module has known bugs and is not maintained well. We cannot
+      recommend using it. You may enable `services.nginx.experimentalZstdSettings`
+      instead which adds the same configuration as the removed option at your own
+      risk.
```
```suggestion
      The zstd module for Nginx has known bugs and is not maintained well.
      It is not generally recommend to use it. You may enable anyway by setting `services.nginx.experimentalZstdSettings`
      which adds the same configuration as the removed option back.
```

Mic92 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -220,7 +220,7 @@ let
       ''}
 
-      ${optionalString cfg.recommendedZstdSettings ''
+      ${optionalString cfg.experimentalZstdSettings ''
```
What other issues does this module has beyond the ordering issue that is solved now?

fpletz replied:
Unfortunately there are still issues remaining that haven't been addressed for a while, like https://github.com/tokers/zstd-nginx-module/pull/23. The last release is from 2023 even though the default branch contains an important bug fix. If we aren't properly maintaining the zstd module ourselves either, we also shouldn't recommend it. I've seen on multiple occasions that people mindlessly turn on all our recommended nginx options. There are still [ordering issues if brotli and zstd are enabled](https://github.com/tokers/zstd-nginx-module/pull/44) which are both recommend by our nginx module.

fpletz committed:
nixos/nginx: remove recommendedZstdSettings, add experimental option

The zstd nginx module has known bugs and upstream is currently not
maintained. We should not recommend a buggy module and configuration
to our users since we are not maintaining the module either.

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -171,6 +171,14 @@ let
           quic_bpf on;
         ''}
 
+        ${optionalString cfg.experimentalZstdSettings ''
```
In line 253 we still have recommendedZstdSettings which will probably break eval and I don't think that the move of the setting was intentional.
